pydmd/bopdmd.py

```diff
@@ -283,6 +292,147 @@ def _varpro_opts_warn(self):
+        imaginary component. Also used to catch an edge case for `omega.imag
+        == 0`.
+
+        :param eigenvalues: Vector of eigenvalues to calculate the difference
+        :param omega: Currently considered eigenvalue
+        :param ind: Index of omega inside of eigenvalues vector
+        :param absolute_diff: Flag dictating difference matrix calculation.
+        :return: Difference vector between omega and other eigenvalues.
+        """
+        if absolute_diff:
+            diff = np.abs(np.abs(eigenvalues.imag) - np.abs(omega.imag))
+            diff[ind] = np.nan
+        else:
+            sign = np.sign(omega.imag)
+            # Catch the edge case of the eigenvalues being exactly 0.
+            if sign == 0.0:
+                diff = np.abs(np.abs(eigenvalues.imag) - omega.imag)
+                diff[ind] = np.nan
+            else:
+                same_sign_index = np.sign(eigenvalues.imag) == sign
+                opp_sign_eigs = np.copy(eigenvalues.imag)
+                opp_sign_eigs[same_sign_index] = np.nan
+                diff = np.abs(np.abs(opp_sign_eigs) - np.abs(omega.imag))
+
+        return diff
+
+    def _force_conj_pairs(self, eigenvalues):
+        """Force pairing of complex eigenvalues.
+
+        :param eigenvalues: Vector of original eigenvalues.
+        :type eigenvalues: numpy.ndarray
+        :return: Vector of forced paired eigenvalues.
+        :rtype: numpy.ndarray
+        """
+
+        num_eigs = len(eigenvalues)
+        new_eigs = np.empty(num_eigs, dtype="complex")
+
+        unassigned_inds = set(np.arange(num_eigs))
+        pair_indices = np.zeros((num_eigs // 2, 2))
+        pair_counter = 0
+        num_expected_pairs = num_eigs
+
+        diff_array = np.zeros((num_eigs, num_eigs))
+
+        # If given an odd number of eigenvalues, find the eigenvalue with
+        # the smallest imaginary part and take it to be a DC mode
+        # eigenvalue.
+        if num_eigs % 2 == 1:
+            ind_0 = np.argmin(np.abs(eigenvalues.imag))
+            new_eigs[ind_0] = eigenvalues[ind_0].real + 0j
+            unassigned_inds.remove(ind_0)
+            eigenvalues[ind_0] = np.nan + 1j * np.nan
+            num_expected_pairs -= 1
+            num_eigs_adj = num_eigs - 1
+        else:
+            num_eigs_adj = num_eigs
+
+        # Determine if we have eigenvalues with opposite signs or if we are
+        # in a weird state where we might not be able to match eigenvalues.
+        if np.count_nonzero(eigenvalues.imag > 0) == np.count_nonzero(
+            eigenvalues.imag < 0
+        ):
+            absolute_diff = False
+        else:
+            absolute_diff = True
+
+        for nomega, omega in enumerate(eigenvalues):
+            # Comparing just the imaginary components allows the
+            # conjugate pair identification to be insensitive to the real
+            # part.
+            diff_array[nomega] = self._diff_func(
```

Comment:
`_diff_func` is a `staticmethod`, thus we should not use `self` to access it

Reply:
@fandreuz I am a little confused because this is how other static methods are accessed throughout PyDMD (e.g., see `_test_kernel_inputs` in `edmd.py` and `compute_irank_svd` in `bopydmd.py`.

How should I access it otherwise?

Reply:
The correct form is `ClassName.static_method_name` :
```suggestion
            diff_array[nomega] = BOPDMDOperator._diff_func(
```

It works because static methods are also accessible by instances of the class